### ISSUE TITLE
export to xstate

### ISSUE THREAD

**Issue by chrisjallen** [sverweij/state-machine-cat#59]
Suggestion to produce xstate output, would be really useful and provide wider audience for use with most popular js library for state charts/ state machines.

Ideally, would be available as an option in the exporter list in the ui side menu.

**Comment by sverweij** [comment]
hi @chrisjallen thanks for this suggestion - it's been an [item](https://github.com/sverweij/state-machine-cat/projects/1#card-12899673) on the backlog for some time as a nice to have feature, but as I don't use xstate and there was no demand (up till now :-) ) it has just stayed there.

That said ...

- In the xstate documentation I read it 'Adheres to the SCXML specification' (third sentence in the link you provided). I haven't dug into xstate deeply, but in theory this'd mean xstate and state-machine-cat are interoperable - at least on some level - already (smcat can emit SCXML). Is this something that can be used?


The item is up in the backlog now, but I'm not sure I'll be able to pick it up in a timely fashion. A PR is welcome though. If xstate is SCXML compliant the [scjson](https://github.com/sverweij/state-machine-cat/tree/develop/src/render/scjson) renderer should provide a good basis.

**Comment by chrisjallen** [comment]
Thanks @sverweij Ill try taking a look at doing this myself, will send back a pr.

**Comment by davidkpiano** [comment]
@sverweij I am working on an `@xstate/scxml` package that will translate XState machine configurations to SCXML (and vice-versa). Would this be enough to get it visualized?

**Comment by sverweij** [comment]
- smcat => scxml => _XState machine configuration_ would then work out of the box 🤞 (this will make @chrisjallen very happy, I gather :-) )
- for XState configuration => scxml => smcat (& visualize it) smcat would need to grow scxml as an input format. This is doable in a reasonable amount of time, though - at least for the scxml _core constructs_...

**Comment by sverweij** [comment]
@davidkpiano couldn't sleep very well last night so I wrote an scxml parser (see PR above). Which means that scxml output should be sufficient to get it visualized.

Available on http://sverweij.gitlab.io/state-machine-cat/ and on npm behind the beta tag; `state-machine-cat@beta` - hope to publish at the start of next week.

(That said - a direct mapping to and from xstate might still be useful ...)

**Comment by sverweij** [comment]
(I've published a polished version to live yesterday)

**Comment by chrisjallen** [comment]
Ah this is great news! I had a go myself with some typescript definitions to boot, but better coming from the authors! Thanks @sverweij @davidkpiano

**Comment by sverweij** [comment]
@chrisjallen - there's definitely value in that, and a PR is still welcome, because likely something will get lost in translation smcat &rlarr; SCXML &rlarr; xstate - because of limitations in SCXML that don't exist in either xstate and/ or in state-machine-cat.

**Comment by chrisjallen** [comment]
Hey @sverweij almost have that pr ready, sorry to pollute this issue but how best to submit pr? No access to push a branch, do I fork?

**Comment by sverweij** [comment]
yep that's how it works; fork, make changes, create a PR. When I made my first PR on another repo I found  GitHub's [creating-a-pull-request-from-a-fork](https://help.github.com/en/articles/creating-a-pull-request-from-a-fork) to be helpful.
